### PR TITLE
V1-163 Creating course endpoint

### PR DESCRIPTION
**Pull request by zhuk-vezunchikz**
Created functionality for creating course on backend side

### CODE REVIEW

File: backend/src/controllers/admin/preparingCourseData.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  req: Request<never, never, ICreateCourseBody>,
+  res: Response,
+  next: NextFunction,
+) => {
+  try {
+    const courseDataFromRequest = req.body;
+
+    const courseDataToSave: ICreateCourseBody = { technologies: [] };
+
+    courseDataToSave.title = courseDataFromRequest.title;
```
Comment (dapodev) [inline backend/src/controllers/admin/preparingCourseData.ts:22]: Validation required

File: backend/src/controllers/admin/preparingCourseData.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    courseDataToSave.complexity = courseDataFromRequest.complexity;
```
Comment (dapodev) [inline backend/src/controllers/admin/preparingCourseData.ts:23]: validation required

File: backend/src/controllers/admin/preparingCourseData.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import { NextFunction, Request, Response } from 'express';
```
Comment (dapodev) [inline backend/src/controllers/admin/preparingCourseData.ts:1]: And move, pls, all the validations to one huge util. And add `throw` to its parts, so to send error to frontend if any validation is not performed successfully.

File: backend/src/controllers/admin/addCourse.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import { addCourseProvider } from 'db/providers/courseProvider';
+import { ICreateCourseBody } from 'interfaces/ICourses/IQueryCourses';
+
+const addCourse = async (
+  req: Request<never, never, ICreateCourseBody>,
+  res: Response<never, { results: ICreateCourseBody }>,
+  next: NextFunction,
+) => {
+  try {
+    const newCourse = res.locals.results;
```
Comment (dapodev) [inline backend/src/controllers/admin/addCourse.ts:12]: We should store as the locals.results value that provider returns after creating new document!

**Review by dapodev — CHANGES_REQUESTED** [review]

File: backend/src/db/providers/skillProvider.ts [hunk 5]
```diff
@@ -194,6 +194,22 @@ const isProperTechnologies = async (techs: IUpdateCourseBody['skills']): Promise
...
+      return SkillModel.findOne({ name: skill });
+    }),
+  );
+
+  const techsForCourse = techs.map((currentSkill, index) => {
+    return {
+      skill: currentSkill?._id as ObjectId,
+      points: technologies[index].points,
+    };
+  });
```
Comment (dapodev) [inline backend/src/db/providers/skillProvider.ts:209]: Make it one-line arrow function pls, if it has only return statement and nothing more

File: backend/src/db/providers/testProvider.ts [hunk 6]
```diff
@@ -88,4 +89,21 @@ const getCourseTest = async (courseId: string | ObjectId): Promise<ITest> => {
...
-export { getTestProvider, getTestById, getTrueAnswersProvider, updateTestQuestions, getCourseTest };
+const addCourseTest = async (testData: ITest) => {
+  const properQuestionsToSet = setAnswerProperNumbersToQuestions(testData.questions);
+
+  return TestModel.create({
+    title: testData.title,
+    questions: properQuestionsToSet,
+    timeout: testData.timeout,
+  });
+};
```
Comment (dapodev) [inline backend/src/db/providers/testProvider.ts:100]: You should only create test here. All the other actions should be done outside provider. Move setting proper numbers to questions outside provider, pls

File: backend/src/utils/validation/isValidComplexity.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import { UserRank } from '../../enums/users';
```
Comment (dapodev) [inline backend/src/utils/validation/isValidComplexity.ts:1]: absolute import

File: backend/src/utils/validation/isValidComplexity.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+
+const isValidComplexity = (complexity?: UserRank): boolean =>
+  Boolean(complexity && typeof complexity === 'number');
```
Comment (dapodev) [inline backend/src/utils/validation/isValidComplexity.ts:4]: not enough checks. complexity also should be positive and should contain only those values, that described in `UserRank` enum. That is really important for some other logic parts of our api

File: backend/src/utils/validation/isValidCourseData.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+const isValidCourseData = (courseData: ICreateCourseBody): boolean => {
+  return (
+    isValidText(courseData.title) &&
+    isValidComplexity(courseData.complexity) &&
+    isValidText(courseData.description) &&
+    isValidMaterials(courseData.materials) &&
+    isValidQuestions(courseData.test?.questions) &&
+    isValidAvatar(courseData.avatar)
+  );
+};
```
Comment (dapodev) [inline backend/src/utils/validation/isValidCourseData.ts:17]: That's nor really good. Pls, call every validation separately and throw an error, that show what actually was wrong. Example: if description is invalid, the message should be "Course description was not validated successfully." or "Invalid course description."

File: frontend/src/api/myCourses/searchClientCourses.ts [hunk 10]
```diff
@@ -7,12 +7,16 @@ import { ClientCourse } from 'types/clientCourse';
...
-  });
+  return useQuery(
+    QUERY_KEYS.myCourses,
+    async () => {
+      const apiClient = apiClientWrapper();
+      const response = await apiClient.get(`${API.getMyCourses}`);
+      const searchResponse = response.data;
+      return searchResponse;
+    },
+    { cacheTime: 900000 },
```
Comment (dapodev) [inline frontend/src/api/myCourses/searchClientCourses.ts:18]: Not related to ticket, but whatever. Magic number is not a good practice.

**Review by dapodev — CHANGES_REQUESTED** [review]

File: backend/src/db/providers/skillProvider.ts [hunk 11]
```diff
@@ -194,6 +194,22 @@ const isProperTechnologies = async (techs: IUpdateCourseBody['skills']): Promise
   return checksPassed;
 };
 
+const skillsToCourseTechs = async (technologies: { skill: ObjectId; points: number }[]) => {
```
Comment (SaXaPhonist) [inline backend/src/db/providers/skillProvider.ts:197]: "getSkillsToCourseTechs"

File: backend/src/db/providers/skillProvider.ts [hunk 12]
```diff
@@ -194,6 +194,22 @@ const isProperTechnologies = async (techs: IUpdateCourseBody['skills']): Promise
...
+  const techs = await Promise.all(
+    technologies.map(({ skill }) => {
+      return SkillModel.findOne({ name: skill });
```
Comment (SaXaPhonist) [inline backend/src/db/providers/skillProvider.ts:200]: it's not quite clear for me how you search by name ( string in model) with { skill } that is equal to Objectid

Comment (SaXaPhonist) [reply backend/src/db/providers/skillProvider.ts:200]: i consider that you can move this promise all out to controller and use instead this provider provider "getCommonSkill"

**Review by SaXaPhonist — CHANGES_REQUESTED** [review]

**Review by dapodev — APPROVED** [review]

**Pull request merged**